### ISSUE TITLE
Mgr pod liveness fails in v6-first dual-stack clusters after #7151

### ISSUE THREAD

**Issue by eyanulis** [rook/rook#7494]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
In an "IPv6-first" dual-stack cluster, the liveness check against the mgr pod consistently fails with "connection refused". This results in the pod repeatedly restarting and eventually going into `CrashLoopBackoff`. This can be worked-around by disabling the liveness check in the `CephCluster` CRD - I haven't found a non-brittle way to reconfigure it such that it can still run.

From `kubectl describe pod`:

```
Events:
  Type     Reason     Age                From               Message
  ----     ------     ----               ----               -------
<normal pull/schedule/start events omitted for brevity>
  Normal   Started    10s (x2 over 99s)  kubelet            Started container mgr
  Warning  Unhealthy  10s (x3 over 30s)  kubelet            Liveness probe failed: Get "http://[2601:282:f00:4c4:4::2af0]:9283/": dial tcp [2601:282:f00:4c4:4::2af0]:9283: connect: connection refused
  Normal   Killing    10s                kubelet            Container mgr failed liveness probe, will be restarted
```
The liveness check is evaluating the Prometheus port and because it is set as `http-get http://:9283/` it will test against the pod IP as reported by Kubernetes. In an IPv6-first dual-stack cluster, the first pod IP returned will be an IPv6 address.

However, in the `init-set-<module>-server-addr` init containers, Prometheus and other services (e.g. the dashboard) are explicitly configured to bind to `0.0.0.0` - an `AF_INET` (IPv4) socket - after #7151 was merged. Thus, a connection to that port on the pod's IPv6 address is refused because there is nothing listening, and the liveness check fails.

A minimal demonstration:
```
[root@rook-ceph-mgr-a-94b5766c6-tf99r /]# ss -tlpn | grep 9283
LISTEN  0       5                           0.0.0.0:9283          0.0.0.0:*     

[root@rook-ceph-mgr-a-94b5766c6-tf99r /]# curl http://127.0.0.1:9283
<!DOCTYPE html>
<html>
    <head><title>Ceph Exporter</title></head>
    <body>
        <h1>Ceph Exporter</h1>
        <p><a href='/metrics'>Metrics</a></p>
    </body>
</html>

[root@rook-ceph-mgr-a-94b5766c6-tf99r /]# curl http://[::]:9283
curl: (7) Failed to connect to :: port 9283: Connection refused
```
I'm not entirely sure _what_ behavior would be seen in an IPv6-only cluster post #7151 since the configured IPv4 bind would likely fail entirely - unfortunately I don't have an easy way to test this at the moment.

**Expected behavior:**
The mgr pod liveness check does not fail in IPv6-first dualstack clusters.

In my opinion (may or may not be correct): Prometheus and the dashboard should be bound to all IPs on the pod (IPv4 and IPv6) - but I'm aware this caused issues for some users e.g. #2526. The [Ceph docs](https://docs.ceph.com/en/latest/mgr/prometheus/#configuration) indicate that by default Prom listens on all IPs, and the user is not explicitly configuring the override to `0.0.0.0` in the Rook config and may not be aware that it is occurring.
The deviation from Ceph's documented default behavior is a tad confusing (took me quite a while to track down where `server_addr` was being set).

**How to reproduce it (minimal and precise):**

Deploy Rook >= 1.5.7 on a "v6-first" dual-stack Kubernetes cluster, and create a basic Ceph cluster. without any additional config (i.e. no storage pools, probably can even get away with zero OSDs).
The `rook-ceph-mgr-a` pod will fail its liveness check and eventually go into `CrashLoopBackoff`.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release): Fedora 33
* Kernel (e.g. `uname -a`): `5.11.8-200.fc33.x86_64 #1 SMP Mon Mar 22 01:35:01 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
* Cloud provider or hardware configuration: Bare metal
* Rook version (use `rook version` inside of a Rook Pod): `v1.5.9`
* Storage backend version (e.g. for ceph do `ceph -v`): `ceph version 15.2.10 (27917a557cca91e4da407489bbaa64ad4352cc02) octopus (stable)`
* Kubernetes version (use `kubectl version`): `v1.20.5`
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): Bare metal configured with kubeadm (dual-stack gates enabled)
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):
```
[root@rook-ceph-tools-6f58686b5d-4cc55 /]# ceph -s
  cluster:
    id:     5ceaa93c-d446-4738-82dc-11bc69559ea6
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum a,b,c (age 95m)
    mgr: a(active, since 47s)
    osd: 27 osds: 27 up (since 93m), 27 in (since 93m)
 
  data:
    pools:   1 pools, 1 pgs
    objects: 27 objects, 0 B
    usage:   27 GiB used, 51 TiB / 51 TiB avail
    pgs:     1 active+clean
```

**Comment by eyanulis** [comment]
(One note: this almost certainly only affects the 1.5 branch, since #7152 removes the init container entirely in master and therefore 1.6.)

**Comment by travisn** [comment]
@eyanulis Did you try setting [ipFamily: v6](https://github.com/rook/rook/blob/release-1.5/cluster/examples/kubernetes/ceph/cluster.yaml#L93)? See [this doc](https://rook.github.io/docs/rook/v1.5/ceph-cluster-crd.html#ipfamily) as well. Not sure that will help with this specific issue, but I thought Ceph didn't support the dual stack.

**Comment by eyanulis** [comment]
Hey @travisn -

I believe you're correct regarding the lack of dual-stack support. Though the environment is dual stack, I'm running the Ceph cluster configured for single-stack IPv6 (including `ipFamily: IPv6` in the cluster CRD and an override of `ms_bind_ipv4 = false` in the `rook-config-override` configmap). With that in place the cluster is working perfectly except for this one issue. As far as I can tell the current configuration of the liveness probe does not take any of the above into consideration.

I've attached the full `CephCluster` CRD (note the liveness check is disabled for the mgr towards the botton).

[cephcluster.txt](https://github.com/rook/rook/files/6233721/cephcluster.txt)

**Comment by leseb** [comment]
> Hey @travisn -
> 
> I believe you're correct regarding the lack of dual-stack support. Though the environment is dual stack, I'm running the Ceph cluster configured for single-stack IPv6 (including `ipFamily: IPv6` in the cluster CRD and an override of `ms_bind_ipv4 = false` in the `rook-config-override` configmap). With that in place the cluster is working perfectly except for this one issue. As far as I can tell the current configuration of the dashboard and Prometheus `server_addr` init containers do not take any of the above into consideration (it is hardcoded to `0.0.0.0` here: https://github.com/rook/rook/blob/release-1.5/pkg/operator/ceph/cluster/mgr/spec.go#L186)
> 
> I've attached the full `CephCluster` CRD (note the liveness check is disabled for the mgr towards the botton).
> 
> [cephcluster.txt](https://github.com/rook/rook/files/6233721/cephcluster.txt)

Ceph's dual stack support will be in the upcoming Pacific release. I was discussing with @sp98 about this and he will followup on adding the missing pieces soon.

**Comment by travisn** [comment]
I believe this is now fixed by #7515 and will be available in v1.6 soon. @leseb Or was some other change needed that you were referring to?

**Comment by leseb** [comment]
> I believe this is now fixed by #7515 and will be available in v1.6 soon. @leseb Or was some other change needed that you were referring to?

Most of it is in  #7515, also the configuration of the `server_addr` disappeared in https://github.com/rook/rook/commit/7dc36e0f4605cc9f3cf7d41f1815998f0e69d2ca. For the liveness probe nothing has changed. 

Also, now that we don't set it anymore (in Rook 1.6), we will leave the mgr module figure out the network stack, see: https://github.com/ceph/ceph/blob/octopus/src/pybind/mgr/mgr_util.py#L108-L124. It seems to behave correctly.

I'm closing this, as I believe everything is in 1.6, and is just around the corner.
Thanks!
